### ISSUE TITLE
Memory Leak in TransitionGroup's onExited

### ISSUE THREAD

**Issue by STRML** [reactjs/react-transition-group#219]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
We're seeing memory leaks with React 16, apparently though the `onExited` handler here in `TransitionGroup`:

https://github.com/reactjs/react-transition-group/blob/8f971b853aef22f3c3b79a919856af5441062729/src/TransitionGroup.js#L164-L176

A screenshot of a heap dump from the leak:
<img width="1648" alt="screen shot 2017-10-04 at 1 50 02 pm" src="https://user-images.githubusercontent.com/1197375/31193970-416b5024-a90b-11e7-9ccd-4926959313a6.png">

A second dump locally shows a reference to the previous child is being held in the closure created by `onExited`:

<img width="1664" alt="screen shot 2017-10-04 at 2 05 32 pm" src="https://user-images.githubusercontent.com/1197375/31194565-2ab3caf8-a90d-11e7-8b31-aecb2ca88591.png">

I'm investigating the leak now, but I figured I'd put this issue in as a reference for a later PR, and to see if anyone else is experiencing this or has any leads.

**Comment by devfive** [comment]
Hi,

I was investigating some problems with this memory leak today and I found this ticket and referenced PR.

https://github.com/reactjs/react-transition-group/commit/06f488729fe864c91c505fc7ae6c27fb6b69097e#diff-ac9d12c3f941c66b21d38abffdfb2885R191
Do we have to bind to handleExited function here? Can't we use `prevChild.props.onExited` which would have handleExited already bound?

I prepared a commit which changes that behaviour and it seems to run much faster on my end (to even 5 times faster on lists with around 50 elements).
https://github.com/devfive/react-transition-group/commit/7aa99e85bb13f4594ffc41fdc5eecb1a743dd8a1

Should I add another issue or will we reopen this?

Cheers,
Mateusz
